Kernel/Storage/NVME/NVMEController.cpp

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2021, Pankaj Raghav <pankydev8@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "NVMEController.h"
+#include "AK/Format.h"
+#include <AK/RefPtr.h>
+#include <AK/Types.h>
+#include <Kernel/Arch/x86/IO.h>
+#include <Kernel/Arch/x86/Processor.h>
+#include <Kernel/Bus/PCI/API.h>
+#include <Kernel/Devices/Device.h>
+#include <Kernel/FileSystem/ProcFS.h>
+#include <Kernel/Sections.h>
+
+namespace Kernel {
+
+NonnullRefPtr<NVMEController> NVMEController::initialize(const Kernel::PCI::DeviceIdentifier& device_identifier)
+{
+    return adopt_ref(*new NVMEController(device_identifier));
+}
+
+NVMEController::NVMEController(const PCI::DeviceIdentifier& device_identifier)
+    : PCI::Device(device_identifier.address())
+    , m_pci_device_id(device_identifier)
+    , m_admin_queue_ready(false)
+    , m_device_count(0)
+{
+    // Nr of queues = one queue per core
```

Comment:
Is this common? Does Linux do something similar, or multiple per core? If not, is this just much simpler to implement?

Reply:
Yeah, Linux also does this. I don't think it is anymore complicated by constructing a queue per core, but the wins are huge. 
Currently the benefits can't be fully utilized in a `smp` system because we are using PIN based interrupts. Once we start supporting MSI interrupts and port NVMe code to use an unique interrupt vector per queue, then it will enable parallel processing without any contention on IO queues.